cmds/ping/ping.go

```diff
@@ -69,47 +77,103 @@ func beatifulLatency(before time.Time) (latency string) {
 	return latency
 }
 
+func cksum(bs []byte) uint16 {
+	sum := uint32(0)
+
+	for k := 0; k < len(bs)/2; k++ {
+		sum += uint32(bs[k*2]) << 8
+		sum += uint32(bs[k*2+1])
+	}
+	if len(bs)%2 != 0 {
+		sum += uint32(bs[len(bs)-1]) << 8
+	}
+	sum = (sum >> 16) + (sum & 0xffff)
+	sum = (sum >> 16) + (sum & 0xffff)
+	if sum == 0xffff {
+		sum = 0
+	}
+
+	return ^uint16(sum)
+}
+
+func ping1(netname string, host string, i uint64) error {
+	c, derr := net.Dial(netname, host)
+	if derr != nil {
+		return fmt.Errorf("net.Dial(%v %v) failed: %v", netname, host, derr)
+	}
+	defer c.Close()
+
+	// Send ICMP Echo Request
+	waitFor := time.Duration(*wtf) * time.Millisecond
+	c.SetDeadline(time.Now().Add(waitFor))
+	msg := make([]byte, *packetSize)
+	msg[0] = ICMP_TYPE_ECHO_REQUEST
+	msg[1] = 0
+	binary.BigEndian.PutUint16(msg[6:], uint16(i))
+	binary.BigEndian.PutUint16(msg[4:], uint16(i>>16))
+	binary.BigEndian.PutUint16(msg[2:], cksum(msg))
+	if _, err := c.Write(msg[:]); err != nil {
+		return fmt.Errorf("Write failed: %v", err)
+	}
+
+	// Get ICMP Echo Reply
+	before := time.Now()
+	c.SetDeadline(time.Now().Add(waitFor))
+	rmsg := make([]byte, *packetSize+256)
+	amt, rerr := c.Read(rmsg[:])
+	if rerr != nil {
+		return fmt.Errorf("Read failed: %v", rerr)
+	}
+	if (rmsg[0] & 0x0F) == 6 {
+		rmsg = rmsg[ICMP_ECHO_RESPONSE_HEADER_IPV6_OFFSET:]
+	} else {
+		rmsg = rmsg[ICMP_ECHO_RESPONSE_HEADER_IPV4_OFFSET:]
+	}
+	if rmsg[0] != ICMP_TYPE_ECHO_RESPONSE {
+		return fmt.Errorf("Bad ICMP echo reply type: %v", msg[0])
+	}
+	cks := binary.BigEndian.Uint16(rmsg[2:])
+	binary.BigEndian.PutUint16(rmsg[2:], 0)
+	if cks != cksum(rmsg) {
+		return fmt.Errorf("Bad ICMP checksum: %v (expected %v)", cks, cksum(rmsg))
+	}
+	latency := beatifulLatency(before)
+	id := binary.BigEndian.Uint16(rmsg[4:])
+	seq := binary.BigEndian.Uint16(rmsg[6:])
+	rseq := uint64(id)<<16 + uint64(seq)
+	if rseq != i {
+		return fmt.Errorf("wrong sequence number %v (expected %v)", rseq, i)
+	}
+	log.Printf("%d bytes from %v: icmp_seq=%v, time=%v", amt, host, i, latency)
```

Comment:
one last thing, I wish I had noticed this, up to you.

The function *could* return string, error, and the caller could print the string if there were no errors.
then this would just return fmt.Sprintf, and writing tests becomes very easy.

Reply:
Done. That's much cleaner.